contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,209 @@
+
+From the events perspective, handling for the following events will be added/updated:
+### Resource Class Creation
+1. Init and add resource class info into local cache
+2. Iterate over all existing nodes in cache to figure out if there are devices on these nodes which are selectable by resource class. If found update the resource class availability status in local cache.
+3. Patch the status of resource class api object with availability state in `ResourceClassStatus`
+
+### Resource Class Deletion
+Delete the resource class info from the cache.
+
+### Node Addition
+Scheduler already caches `NodeInfo`. Now additionally update device state:
+1. Check in the node status if any devices are present.
+2. For each device found, iterate over all existing resource classes in the cache to find resource classes which can select this particular device.
+3. 'allocatable' of the resource class will be computed and updated in local cache.
+4. ResourceClass api object's status will be patched with the new “allocatable” value
+
+### Node Deletion
+If node has devices which are selectable by existing resource classes:
+1. Adjust resource class state in local cache.
+2. Update resource class status by patching api object.
+
+### Pod Creation
+1. Get the requested resource class name and quantity from pod spec.
+2. Select nodes by applying predicates according to requested quantity and Resource class's state present in the cache.
+3. On the selected node, select a Device from the stored devices info in cache after matching key,value from requested resource class.
+4. After device selection, update(decrease) “Requested” for all the resource classes which could select this device in the cache.
+5. Patch the resource class objects with new “Requested” in the `ResourceClassStatus`.
+6. Add the pod reference in local DeviceToPod mapping structure in cache
+7. Patch the pod object with selected device annotation with prefix 'scheduler.alpha.kubernetes.io/resClass'
+
+### Pod Delete
+1. Iterate over the all the devices on the node that this pod was scheduled to and find out the devices being used by pod.
+2. For each device consumed by pod, update availability state of Resource classes which can select this device in the cache.
+3. Patch `ResourceClassStatus` with new availability state.
+
+## Kubelet Changes
+Update logic at container runtime manager to look for device annotations, prefixed by 'scheduler.alpha.kubernetes.io/resClass' and call matching device plugins.
+
+## Future Scope
+* RBAC: It can further be explored that how to tie resource classes with RBAC like any other existing API resource objects.
+* Nested Resource Classes: In future device plugins and resource classes can be extended to support the nested resource class functionality where one resource class could be comprised of a group of sub-resource classes. For example 'numa-node' resource class comprised of sub-resource classes, 'single-core'.
```

Comment:
We need to communicate the plans/decisions on how ResourceClasses relate to Opaque Integer Resources.

@ConnorDoyle